File: backend/puzzles/engine/games/sudoku.py
from enum import IntEnum
from typing import TYPE_CHECKING

from puzzles.abstract import PuzzleDefinition
from puzzles.engine.games.base import PuzzleEngineBase, State
from puzzles.engine.handlers.sudoku_input import SudokuInputHandler
from puzzles.engine.rules.sudoku import no_duplicate_numbers_in_rows, no_duplicate_numbers_in_cols, \
    no_duplicate_numbers_in_boxes

if TYPE_CHECKING:
    pass

class SudokuCellStates(IntEnum):
    Empty = 0
    Filled = 1
    Crossed = 2


class SudokuEngine(PuzzleEngineBase):
    def __init__(self, definition: PuzzleDefinition, board_state: State) -> None:
        super().__init__(definition, board_state, input_handler=SudokuInputHandler(), validation_constraints=[
            no_duplicate_numbers_in_rows(),
            no_duplicate_numbers_in_cols(),
            no_duplicate_numbers_in_boxes(),
        ])

    def is_solved(self, strict=False) -> bool:
        """
        Check if the Sudoku puzzle is solved.
        :param strict: Ignored. Sudoku is inherently strict.
        :return:
        """
        board = "".join([str(c) for c in self.board_state])
        solution = "".join([str(c) for c in self.get_solution_board_string()])
        return board == solution

    def can_modify_cell(self, _state: State, row: int, col: int) -> bool:
        initial_board = self.get_initial_board_string()

        return initial_board[row * self.cols + col] == 0

    def calculate_number_of_mistakes(self) -> int:
        """
        Calculate the number of mistakes in the current board state.
        :return: Number of mistakes.
        """
        solution = self.get_solution_board_string()
        mistakes = 0
        for row in range(self.rows):
            for col in range(self.cols):
                if not self.can_modify_cell(self.board_state, row, col):
                    continue
                if self.board_state[row * self.cols + col] != solution[row * self.cols + col]:
                        mistakes += 1
        return mistakes

File: frontend/src/services/game/engines/games/sudoku.ts
import { type InputHandler, PuzzleEngine } from "@/services/game/engines/PuzzleEngine.ts";
import type { Cell } from "@/features/games.components/board.interaction.ts";
import type { PuzzleDefinition } from "@/services/game/engines/types.ts";
import { SudokuCell } from "@/services/game/engines/translator.ts";
import {
  no_duplicate_numbers_in_boxes,
  no_duplicate_numbers_in_cols,
  no_duplicate_numbers_in_rows,
} from "@/services/game/engines/rules/sudoku.ts";

export class SudokuInputHandler implements Partial<InputHandler> {
  on_cell_key_down(engine: PuzzleEngine, cell: Cell, key: string): boolean {
    if (cell.zone !== "game") return false;

    // Handle backspace/delete to clear cell
    if (key === "Backspace" || key === "Delete") {
      engine.board_state[cell.row][cell.col] = 0;
      return true;
    }

    // Handle numeric input
    const num = parseInt(key);
    if (isNaN(num) || num < 1 || num > engine.definition.rows) return false;
    engine.board_state[cell.row][cell.col] = num;
    return true;
  }
}

export class EngineSudoku<T> extends PuzzleEngine<T> {
  constructor(definition: PuzzleDefinition<T>, board_state?: number[][]) {
    super(
      definition,
      board_state,
      [new SudokuInputHandler()],
      [no_duplicate_numbers_in_rows(), no_duplicate_numbers_in_cols(), no_duplicate_numbers_in_boxes()],
    );
  }

  can_modify_cell = (cell: Cell): boolean => this.initial_state[cell.row][cell.col] === SudokuCell.EMPTY;
}
